plugins/modules/mso_schema_template_anp_epg_annotation.py

```diff
@@ -0,0 +1,229 @@
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    schema: Schema 1
+    template: Template 1
+    anp: ANP 1
+    epg: EPG 1
+    state: query
+  register: query_all
+
+- name: Delete an annotation
+  cisco.mso.mso_schema_template_anp_epg_annotation:
+    host: mso_host
+    username: admin
+    password: SomeSecretPassword
+    schema: Schema 1
+    template: Template 1
+    anp: ANP 1
+    epg: EPG 1
+    annotation_key: annotation_key_1
+    state: absent
+"""
+
+RETURN = r"""
+"""
+
+import copy
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.mso.plugins.module_utils.mso import MSOModule, mso_argument_spec
+from ansible_collections.cisco.mso.plugins.module_utils.schema import MSOSchema
+
+
+def main():
+    argument_spec = mso_argument_spec()
+    argument_spec.update(
+        schema=dict(type="str", required=True),
+        template=dict(type="str", required=True),
+        anp=dict(type="str", required=True),
+        epg=dict(type="str", required=True),
+        annotation_key=dict(type="str", aliases=["key"], no_log=False),
+        annotation_value=dict(type="str", aliases=["value"]),
+        state=dict(type="str", default="present", choices=["absent", "present", "query"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "absent", ["annotation_key"]],
+            ["state", "present", ["annotation_key", "annotation_value"]],
+        ],
+    )
+
+    schema = module.params.get("schema")
+    template = module.params.get("template")
+    anp = module.params.get("anp")
+    epg = module.params.get("epg")
+    annotation_key = module.params.get("annotation_key")
+    annotation_value = module.params.get("annotation_value")
+    state = module.params.get("state")
+
+    mso = MSOModule(module)
+
+    mso_schema = MSOSchema(mso, schema, template)
+    mso_schema.set_template(template)
+    mso_schema.set_template_anp(anp)
+    mso_schema.set_template_anp_epg(epg)
+
+    if annotation_key:
+        mso_schema.set_template_anp_epg_annotation(annotation_key, False)
+        annotation = mso_schema.schema_objects.get("template_anp_epg_annotation")
+        if annotation is not None:
+            mso.existing = mso.previous = copy.deepcopy(annotation.details)  # Query a specific Annotation
+    else:
+        epg_object = mso_schema.schema_objects["template_anp_epg"]
+        mso.existing = epg_object.details.get("tagAnnotations", [])  # Query all
+
+    path = "/templates/{0}/anps/{1}/epgs/{2}/tagAnnotations".format(template, anp, epg)
+
+    ops = []
+    if state == "present":
+        mso_values = {"tagKey": annotation_key, "tagValue": annotation_value}
+        mso.sanitize(mso_values)
+        if mso.existing:
+            # An error from MSO prevents the "replace" operation from updating existing annotation values.
```

Comment:
Just out of curiosity, what errors do you get from this exactly? on which MSO versions?

Reply:
`MSO Error 400: json: cannot unmarshal string into Go struct field Schema2.templates of type common.Epg2`
Version 4.3.1 - This is the only version I have access to right now.
However, I have found a way to use the replace op instead.